### PR TITLE
Tt2 942 gdpr add support for daterange for data retrieval

### PR DESCRIPTION
**Pull request by zaydgds**
- add cli command `retrieve-audit-data` to start glacier restore and double decryption
  - add unit tests too 
- repo organisation to so each cli command gets it own subfolder
- eslint updated to support types for `testVariadicArgs` in `scripts/utils/cliUtils.ts` 

out of scope work that was done
- organising the cli command implementation into individual folders
- updated tests for all the cli commands

### CODE REVIEW

**Comment by katiewhitlock** [comment]
Hi Zayd,

It's crucial to maintain a close alignment between the scope of the ticket and the changes made in the PR. This helps us keep our codebase clean, ensures that we have clear documentation of what changes were made, and makes it easier for everyone to understand the purpose of each PR - not just for our current team but those who will be maintaining this way into the future. 

This PR does way more than outlined in the ticket [here](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&selectedIssue=TT2-942), and overlaps with CLI work which was documented for this ticket [here.](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&assignee=712020%3A66f429f7-ea49-4593-ab6b-8628100b6c3f&selectedIssue=TT2-984). 

Please either:
a) Write well documented descriptions of ALL the changes you are making unrelated to the ticket, and why.
b) Resubmit this with just the changes related to ticket TT2-942.

Thanks.

**Comment by zaydgds** [comment]
> Hi Zayd,
> 
> It's crucial to maintain a close alignment between the scope of the ticket and the changes made in the PR. This helps us keep our codebase clean, ensures that we have clear documentation of what changes were made, and makes it easier for everyone to understand the purpose of each PR - not just for our current team but those who will be maintaining this way into the future.
> 
> This PR does more than outlined in the ticket [here](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&selectedIssue=TT2-942), and overlaps with CLI work which was documented for this ticket [here.](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&assignee=712020%3A66f429f7-ea49-4593-ab6b-8628100b6c3f&selectedIssue=TT2-984).
> 
> Please either: a) Write well documented descriptions of ALL the changes you are making unrelated to the ticket, and why. b) Resubmit this with just the changes related to ticket TT2-942, and add other changes to separate tickets.
> 
> Thanks.



> Hi Zayd,
> 
> It's crucial to maintain a close alignment between the scope of the ticket and the changes made in the PR. This helps us keep our codebase clean, ensures that we have clear documentation of what changes were made, and makes it easier for everyone to understand the purpose of each PR - not just for our current team but those who will be maintaining this way into the future.
> 
> This PR does more than outlined in the ticket [here](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&selectedIssue=TT2-942), and overlaps with CLI work which was documented for this ticket [here.](https://govukverify.atlassian.net/jira/software/c/projects/TT2/boards/238?assignee=631f298e4395a525a710769a&assignee=712020%3A66f429f7-ea49-4593-ab6b-8628100b6c3f&selectedIssue=TT2-984).
> 
> Please either: a) Write well documented descriptions of ALL the changes you are making unrelated to the ticket, and why. b) Resubmit this with just the changes related to ticket TT2-942, and add other changes to separate tickets.
> 
> Thanks.

curious to know what you feel went out of scope of the ticket? Any thing in particular?

File: package.json [hunk 1]
```diff
@@ -21,6 +21,7 @@
     "setupDevStackSecrets": "ts-node scripts/setupDevStackSecrets.ts",
     "test": "jest -c jest.config.ts",
     "test:cov": "jest -c jest.config.ts --coverage",
+    "test:cov:watch": "concurrently \"python3 -m http.server --directory \"coverage/lcov-report\"\" \"yarn test --coverage --watchAll\"",
```
Comment (zaydgds) [inline package.json:24]: reverting due to https://govukverify.atlassian.net/browse/TT2-1069

File: scripts/utils/lib.ts.ts
Comment (dannymadden-gds) [inline scripts/utils/lib.ts.ts:1]: reminder to rename this to lib.ts

File: README.md [hunk 2]
```diff
@@ -200,12 +200,13 @@ yarn lint
...
 All scripts can now be ran using `yarn cli` in the terminal. Use `yarn cli --help` to see what scripts are available and how to use them. You can also run `yarn cli <command> --help` to view detailed help per command
 
-`script/cli.ts` is the entrypoint to the cli, each command listed by `yarn cli --help` will be implemented under `scripts/{command}/`. each command should have detailed guidance on what the command does and details on the mandatory **arguments** and the **optional** options that can be provided to the CLI.
+`script/cli.ts` is the entrypoint to the cli, each command listed by `yarn cli --help` will be implemented under `scripts/{command}/`. each command should have detailed guidance in `--help` on what the command does and details on the mandatory **arguments** and the **optional** options that can be provided to the CLI.
 
 > [!NOTE]  
 > Not all scripts have been migrated over. The following scripts can be used
 >
 > - Sending results of raw audit data
+> - retireving raw audit data from glacier and/or double encryption
```
Comment (dannymadden-gds) [inline README.md:209]: spelling

File: scripts/utils/dateUtils.test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  { date: '2020/01/01-2020/01/9', expected: false },
+  { date: '2020/01/01-2020/1/09', expected: false },
+  { date: '2020/01/01-2020/1/9', expected: false },
+  { date: '2020/01/01-2020/01-09', expected: false },
+  { date: '2020/01/01-2020-01/09', expected: false },
+  { date: '2020/01/01-2020-01-09', expected: false },
+  { date: '2020/01-01-2020/01/09', expected: false },
+  { date: '2020-01/01-2020/01/09', expected: false },
+  { date: '2020-01-01-2020/01/09', expected: false }
+]
```
Comment (dannymadden-gds) [inline scripts/utils/dateUtils.test.ts:53]: is it worthwhile adding '-' in here for no values provided for date range? happy to be told not needed

File: scripts/cli.ts [hunk 4]
```diff
@@ -1,40 +1,99 @@
...
+          recipientName: recipientName,
+          zendeskId: zendeskId
+        }).then(() => {})
+      }
+    )
+
+  program
+    .command('retrieve-audit-data')
+    .description(
+      'Hooks into SAL to retrieve audit data from glaicer and perform double decryption of audit data so it can be queried by athena'
```
Comment (dannymadden-gds) [inline scripts/cli.ts:56]: spelling glaicer -> glacier

File: scripts/initiateCopyAndDecrypt/manualAuditDataRequestInitiateCopyAndDecryptAction.ts
Comment (dannymadden-gds) [inline scripts/initiateCopyAndDecrypt/manualAuditDataRequestInitiateCopyAndDecryptAction.ts:1]: does this file need a test?

File: scripts/utils/cliUtils.test.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        '123',
+        sampleTesterFunctionIsString
+      )
+    } catch (error) {
+      expect(error).toBeInstanceOf(InvalidArgumentError)
+      expect((error as Error).message).toBe(
+        'Value provided does not match supported format'
+      )
+    }
+  })
```
Comment (katiewhitlock) [inline scripts/utils/cliUtils.test.ts:22]: This is difficult to read/requires high mental load to unpick. 

Please extract the three arguments into clearly named variables e.g. 
    const invalidNumberAsString = 123 as unknown as string;
    const validString = '123';
    const invalidFunction = sampleTesterFunctionIsString

    try {
      testVariadicArgs(invalidNumberAsString, validString, invalidFunction); } etc...

File: scripts/utils/cliUtils.ts [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import { InvalidArgumentError } from '@commander-js/extra-typings'
+
+export const testVariadicArgs = (
+  currentValue: string,
+  previousValue: string | string[],
```
Comment (katiewhitlock) [inline scripts/utils/cliUtils.ts:5]: why might we expect the previousValue be an array of strings and not the currentValue?

File: scripts/utils/cliUtils.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { InvalidArgumentError } from '@commander-js/extra-typings'
+
+export const testVariadicArgs = (
```
Comment (katiewhitlock) [inline scripts/utils/cliUtils.ts:3]: Nitpick: are we likely to have any more variadic arguments in this repo? I'd be inclined just to use standard date validation as this seems a layer of complexity we don't need? Appreciate it is elegant though and nice reusability.

File: scripts/utils/dateUtils.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
+import { testVariadicArgs } from './cliUtils'
+
+export const isDateString = (date: string): boolean => {
+  const dateFormatRegex = /^\d{4}-\d{2}-\d{2}$/
```
Comment (katiewhitlock) [inline scripts/utils/dateUtils.ts:4]: Nitpick: it's good practice to place this in a constants.ts file and then use it in here and the tests

File: scripts/utils/dateUtils.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+export const testDateArgs = (
+  currentValue: string,
+  previousValue: string | string[]
+) => {
+  return testVariadicArgs(currentValue, previousValue, isDateString)
+}
+
+export const isDateRange = (daterange: string) => {
+  const daterangeRegex = /^\d{4}\/\d{2}\/\d{2}-\d{4}\/\d{2}\/\d{2}$/
```
Comment (katiewhitlock) [inline scripts/utils/dateUtils.ts:17]: This too

File: scripts/utils/dateUtils.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+export const testDateRangeArgs = (
+  currentValue: string,
+  previousValue: string | string[]
+) => {
+  return testVariadicArgs(currentValue, previousValue, isDateRange)
+}
+
+export const convertDaterangeToDateArray = (daterangeArray: string[]) => {
```
Comment (katiewhitlock) [inline scripts/utils/dateUtils.ts:28]: Ripe for a JSDoc this one.

**Review by dannymadden-gds — APPROVED** [review]

**Review by katiewhitlock — APPROVED** [review]

**Pull request merged**